quantum/rgb_matrix/rgb_matrix.c

```diff
@@ -426,12 +428,31 @@ void rgb_matrix_task(void) {
 
     uint8_t effect = suspend_backlight || !rgb_matrix_config.enable ? 0 : rgb_matrix_config.mode;
 
+    // only for static modes or upon initialisation
+    if (( // static animation always enabled
+#ifdef ENABLE_RGB_MATRIX_ALPHAS_MODS
+        rgb_matrix_config.mode != RGB_MATRIX_ALPHAS_MODS &&
+#endif
+#ifdef ENABLE_RGB_MATRIX_GRADIENT_UP_DOWN
+        rgb_matrix_config.mode != RGB_MATRIX_GRADIENT_UP_DOWN &&
+#endif
+#ifdef ENABLE_RGB_MATRIX_GRADIENT_LEFT_RIGHT
+        rgb_matrix_config.mode != RGB_MATRIX_GRADIENT_LEFT_RIGHT &&
+#endif
+    rgb_matrix_config.mode != 1 ) || rgb_effect_params.init) {
+        render_update = true;
+    }
+
     switch (rgb_task_state) {
         case STARTING:
             rgb_task_start();
+            render_update = true;
             break;
         case RENDERING:
-            rgb_task_render(effect);
+            if (render_update) {
+                rgb_task_render(effect);
+                render_update = false;
+            }
```

Comment:
Unfortunately, skipping the `rgb_task_render()` call like this would probably break the existing implementations of `rgb_matrix_indicators_user()` and `rgb_matrix_indicators_advanced_user()` that use “transparent” indicators (override the LED state only when the indicator is on) — the original LED state for such indicators will never be restored. I'm not sure how to fix this problem without major API changes (and maybe the resulting code would actually be more expensive than just refreshing the static effects every frame, especially because the actual LED updates would need to be done anyway because of indicators).

Reply:
Hmmm you're right. Would it be better to have a return `true` or `false` in the rgb_indicators when there is a change so that it can be rendered?

Reply:
I think another approach here could be to use a "dirty" flag on the actual LED state array that holds the RGB values per LED. If `set_led` is called for a specific index, and the new value is different from the old, it becomes dirty -- and flushing out the data via the ws2812 driver then clears the dirty bit and won't flush again unless it's dirty again.

I tried to implement this, but it actually had a minor negative effect on my matrix scan rate - I believe all the branching and if-checks are slower (on rp2040/arm) than a straight-through write call. YMMV.